File: singly_linked_list.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    insert_at_beginning(data) {
        const new_node  = new Node(data);
        if (this.head === null) {
            this.head = new_node ;
        } else {
            new_node.next = this.head;
            this.head = new_node ;
        }
    }

    insert_in_between(data, key) {
        const new_node  = new Node(data);
        let temp = this.head;
        if (this.head === null) {
            this.head = new_node ;
            return;
        }
        while (temp !== null) {
            if (temp.data === key) {
                new_node .next = temp.next;
                temp.next = new_node ;
                return;
            }
            temp = temp.next;
        }
        return -1;
    }

    insert_at_end(data) {
        const new_node  = new Node(data);
        if (this.head === null) {
            this.head = new_node ;
            return;
        }
        let temp = this.head;
        while (temp.next !== null) {
            temp = temp.next;
        }
        temp.next = new_node ;
    }

    delete_node(key) {
        let prev = null;
        let temp = this.head;
        if (temp !== null && temp.data === key) {
            this.head = temp.next;
            return;
        }
        while (temp.next !== null) {
            if (temp.data === key) {
                break;
            }
            prev = temp;
            temp = temp.next;
        }
        if (temp.next === null) {
            if (temp.data === key) {
                prev.next = null;
                return;
            }
            return -1;
        }
        prev.next = temp.next;
    }

    print_linked_list() {
        let temp = this.head;
        while (temp !== null) {
            console.log(temp.data + " ");
            temp = temp.next;
        }
        console.log();
    }
}

const linkedList = new LinkedList();
linkedList.insert_at_beginning(20);
linkedList.insert_at_beginning(25);
linkedList.insert_at_end(45);
linkedList.insert_in_between(67, 20);
linkedList.delete_node(20);
linkedList.print_linked_list();

File: singly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    def insert_at_beginning(self, data):
        new_node=Node(data)
        if self.head is None:
            self.head = new_node
            return 0
        new_node.next = self.head
        self.head = new_node
        return 0
    def insert_in_between(self,data,key):
        new_node=Node(data)
        temp = self.head
        if self.head is None:
            self.head = new_node
            return 0
        while temp is not None:
            if temp.data==key:
                new_node.next=temp.next
                temp.next=new_node
                return 0
            temp = temp.next
        return -1
    
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return 0
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
        return -1
    def delete_node(self,key):     
        prev=None;
        temp=self.head;
        if temp is not None and temp.data==key:
            self.head=temp.next;
            del temp
            return 0;
        while temp.next!=None:
            if temp.data==key:
                break;
            prev=temp;
            temp=temp.next;
        if (temp.next is None):
            if (temp.data==key): # to deleate last element if key was at last
                prev.next=None;
                del temp
                return 0;
            
            return -1;
        prev.next=temp.next;
        del temp;
        return 0;
    def print_linked_list(self):
        temp = self.head
        while temp is not None:
            print(temp.data, end=" ")
            temp = temp.next
        print()

# Example usage:

linked_list = LinkedList()
linked_list.insert_at_beginning(20)
linked_list.insert_at_beginning(25)
linked_list.insert_at_end(45)
linked_list.insert_in_between(67,20)
linked_list.delete_node(20)
linked_list.print_linked_list()

